Python/specialize.c

```diff
@@ -778,6 +789,100 @@ _Py_Specialize_StoreAttr(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name, S
 }
 
 
+int
+_Py_Specialize_LoadMethod(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name, SpecializedCacheEntry *cache)
+{
+    _PyAdaptiveEntry *cache0 = &cache->adaptive;
+    _PyAttrCache *cache1 = &cache[-1].attr;
+    _PyObjectCache *cache2 = &cache[-2].obj;
+
+    PyTypeObject *owner_cls = Py_TYPE(owner);
+    if (PyModule_CheckExact(owner)) {
+        // Mabe TODO: LOAD_METHOD_MODULE
+        SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_MODULE_METHOD);
+        goto fail;
+    }
+    if (owner_cls->tp_dict == NULL) {
+        if (PyType_Ready(owner_cls) < 0) {
+            return -1;
+        }
+    }
+    PyObject *descr = NULL;
+    DesciptorClassification kind = analyze_descriptor(owner_cls, name, &descr, 0);
```

Comment:
You need to store the version here, otherwise there is a risk that it changes and the type version no longer protects `descr` from having been freed. We need to be careful when using borrowed references.

I think you are correct in using borrowed references. It will save a lot of overhead when freeing the code object, and prevents leaks. But we do need to be extra careful.

Reply:
Ah this is a really good catch! Thanks, I don't think I would've noticed that myself.

Indeed we're on a fine line between performance and safety. The borrowed method is slightly faster than the load_attr hint-style, and most importantly it works for inherited methods -- hints don't work unless we also store the index for the object in the MRO the hint belongs to. But that just makes complexity balloon (it's also slightly slower).